Rewriting of the registration mechanism

RedTachyon opened the pull request:
# Description

This is a total rewrite of `gym.envs.regitration` as proposed in #2738.

It technically doesn't really add any new features beyond fixing a few things that used not to work, but they weren't necessarily documented, so it's hard to call it a bug fix either.

Instead, this significantly simplifies the implementation of registration so that we can actually start adding features to it

The current version implements all significant* behaviors from the old registration mechanism. The intended mechanism is as follows:

1. The registry is a global dictionary sitting in `gym.envs.registration`
2. When registering an environment, you create `EnvSpec` and add it to the registry
3. When creating an environment, you get the appropriate `EnvSpec` from the registry
4. If the passed environment name is invalid, try to give an extra hint on what went wrong ("You tried to create "cartploe", did you mean "CartPole"?)
5. If the passed environment name does not specify the version, create the newest version
6. There is a support for namespaces in the format `namespace/EnvName-vX` (this already existed, but was undocumented, so I'll add it to the docs somewhere later)

One thing I have yet to add is adding an optional minimal env checker on `make` to check if the environment uses the up-to-date API.

*Some low-level details (like the EnvSpecTree) are removed. Also, the error messages may be a bit different, but following the same spirit, and passing the previous tests for where errors and warnings are expected. This is with the exception of one test which was just bugged.

## Highlights

To justify this change, here's a comparison of how `make` changed

In the old implementation, you have `gym.make` which takes an environment name and any kwargs, and instantly just calls `registry.make`. This finds the right `EnvSpec`, and then calls `spec.make`. The actual env creation logic lives in `spec.make`.

Now, it's all handled by `gym.make` -- it finds the right `spec` in the `registry`, initializes it, applies whatever wrappers necessary, and returns the environment.

This is significant when trying to change something about `gym.make`. Previously, you'd have to propagate any changes across the three functions. Now you just change the one function responsible for that. 

An extra bug that this accidentally solves is something that lived in the `register` function. In the old code, all of its logic was in a `try-except-finally` block, where the actual registration happened in `finally`. So even if you completely malformed your environment registration, it will throw an exception and then... still register it. 

Notably, `EnvSpecTree` has been removed and replaced with a dictionary and a few functions. This might cause a worse asymptotical lookup complexity if someone has thousands of environments and namespaces registered, and tries to create an incorrectly named environment, but that's about it (if you have the right environment name, the lookup is still just a dictionary lookup)

## Note:
This is a very core part of the code, so will require significant review. I'm going for full backwards compatibility in terms of `make`, `register` and `spec`.


# Checklist:

- [x] I have run the [`pre-commit` checks](https://pre-commit.com/) with `pre-commit run --all-files` (see `CONTRIBUTING.md` instructions to set it up)
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

pseudo-rnd-thoughts commented:
Thanks for the changes, I can't see any obvious issues with the code though I haven't done a thorough review
You say that you want to preserve full backward compatibility of the core feature but some of the low-level features have been removed. 
Can you think of a reasonable case where removing the low-level features could break someone's code? I.e., should this be released as part of 1.0 or can this be pre 1.0?

RedTachyon commented:
I don't think there are *reasonable* cases where this would break someone code. The removed features are undocumented implementation details and weren't intended for end-user usage as far as I know.

But you [never know](https://xkcd.com/1172/)

ikamensh commented on tests/envs/test_registration.py:
```diff
@@ -188,112 +193,108 @@ def test_spec_with_kwargs():
+    register(id="Test1-v15", entry_point=None)
+    register(id="Test1-v9", entry_point=None)
+    register(id="Other1-v100", entry_point=None)
     try:
```
is this construct a poor man's version of `with pytest.raises(error.DeprecatedEnv)` ?

RedTachyon replied:
I think so, I wanted to make as few changes to the tests as possible, but might as well make some updates to them as well

RedTachyon replied:
I'll keep this out of the scope of this PR for now, this was taken from the original code, and this PR is meant to primarily improve a specific aspect of registration, and keep the functionality otherwise. This can be updated later

ikamensh commented on gym/envs/registration.py:
```diff
@@ -680,63 +310,168 @@ def make(id: Literal[
+current_namespace: Optional[str] = None
+
+
+def check_spec_register(spec: EnvSpec):
```
This public function lacks a docstring. What it means to "check" spec register?

RedTachyon replied:
Good catch, I'll add that

RedTachyon replied:
This should probably be renamed. The intention is that it performs a bunch of checks to make sure that the registration will be valid. It also shouldn't be public

ikamensh commented on gym/envs/registration.py:
```diff
@@ -680,63 +310,168 @@ def make(id: Literal[
+    current_namespace = old_namespace
+
+
+def register(id: str, **kwargs):
```
public function w/o docstring

RedTachyon replied:
Good catch, I'll add that (and the others)

ikamensh commented on gym/envs/registration.py:
```diff
@@ -680,63 +310,168 @@ def make(id: Literal[
+    registry[spec.id] = spec
+
+
+def make(
```
public function w/o docstring

ikamensh commented on gym/envs/registration.py:
```diff
@@ -680,63 +310,168 @@ def make(id: Literal[
+    return env
+
+
+def spec(env_id: str) -> EnvSpec:
```
public function w/o docstring

ikamensh commented on tests/envs/test_registration.py:
```diff
@@ -188,112 +193,108 @@ def test_spec_with_kwargs():
+    spec("test/Test3")
+
+
+# def test_env_spec_tree():
```
Since `EnvSpecTree` class is deleted, it's probably best to delete this test code as well, instead of commenting out.

ikamensh reviewed (commented):
Left some generic software quality comments. Re: docstrings, even a one-liner re-affirming what you would guess from the name can help reader be confident they didn't misunderstand the purpose of a method.

In general I'm also a fan of adding one-liner comments to unit test to say what it's intended to verify, but this is not a core contribution of this PR anyway.

younik reviewed (approved):
Well done! LGTM

pseudo-rnd-thoughts commented on gym/envs/registration.py:
```diff
@@ -7,34 +7,29 @@
-from gym import Env, error, logger
-from gym.envs.__relocated__ import internal_env_relocation_map
-
 if sys.version_info >= (3, 8):
```
As we build for python 3.7 and greater then this is not needed anymore

pseudo-rnd-thoughts replied:
I tested for python 3.7 and without the `__future__` then the import fails however with the `__future__` I didnt have an issue

pseudo-rnd-thoughts commented on gym/envs/registration.py:
```diff
@@ -78,558 +75,191 @@ def parse_env_id(id: str) -> Tuple[Optional[str], str, Optional[int]]:
-
-        if self.order_enforce:
-            from gym.wrappers.order_enforcing import OrderEnforcing
+def check_namespace_exists(ns: Optional[str]):
```
docstring just for the future if someone is trying to understand the function

pseudo-rnd-thoughts commented on gym/envs/registration.py:
```diff
@@ -78,558 +75,191 @@ def parse_env_id(id: str) -> Tuple[Optional[str], str, Optional[int]]:
 
-        if self.autoreset:
-            from gym.wrappers.autoreset import AutoResetWrapper
+def check_name_exists(ns: Optional[str], name: str):
```
docstring

pseudo-rnd-thoughts commented on gym/envs/registration.py:
```diff
@@ -78,558 +75,191 @@ def parse_env_id(id: str) -> Tuple[Optional[str], str, Optional[int]]:
-                message += f" Did you mean: `{suggestions[0]}`?"
-        # Throw the error
-        raise error.NameNotFound(message)
+def check_version_exists(ns: Optional[str], name: str, version: Optional[int]):
```
docstring

pseudo-rnd-thoughts commented on gym/envs/registration.py:
```diff
@@ -78,558 +75,191 @@ def parse_env_id(id: str) -> Tuple[Optional[str], str, Optional[int]]:
+                module, attr = plugin.value.split(":", maxsplit=1)
+            else:
+                module, attr = plugin.value, None
+        except:
```
Would we want to raise a warning here to say there was an issue loading a plugin, otherwise a plugin may not be loaded for unknown reasons?

RedTachyon replied:
I'd need to be sure about the consequences of that. This code is just copy-paste of the old implementation, so I'll keep this outside the scope of this PR

pseudo-rnd-thoughts commented on gym/envs/registration.py:
```diff
@@ -78,558 +75,191 @@ def parse_env_id(id: str) -> Tuple[Optional[str], str, Optional[int]]:
-                )
-            # Replace namespace
-            spec.namespace = self._ns
+def find_newest_version(ns: Optional[str], name: str) -> Optional[int]:
```
docstring

pseudo-rnd-thoughts replied:
Very minor point, would `find_highest_version` be a more accurate name? Newest is possibly confusing as you could think the latest spec to be added

RedTachyon replied:
Renamed

pseudo-rnd-thoughts commented on gym/envs/registration.py:
```diff
@@ -680,63 +310,190 @@ def make(id: Literal[
+        (
+            spec_
+            for spec_ in registry.values()
+            if spec_.name == spec.name and spec_.version is not None
```
Do we need to check that the `spec_.namespace == spec.namespace` as there may be two environments, one with a namespace and the other without. i.e. atari where we have `ALE/Pong-v5` and `Pong-v4` but if the versions were the same then could this be an issue?

RedTachyon replied:
Good catch, fixed

pseudo-rnd-thoughts commented on tests/envs/test_registration.py:
```diff
@@ -2,9 +2,9 @@
 import gym
 from gym import envs, error
-from gym.envs import registration
+from gym.envs import register, registration, registry, spec
```
This shouldn't matter in implementation but in this test you sometimes do `envs.register` and sometimes `register` which got me confused. Could you only one of the method to make it consistent as I was getting confused thinking that they were different

RedTachyon replied:
Fixed

arjun-kg commented:
Registering the exact same ID again now raises an error as opposed to overriding with a warning before. Is this intentional? (Would this be in issue for backward compatibility or would noone ever intentionally do this?)

Also, couple of things not related to the PR changes, but part of registration - 
1. The malformed env error is not human readable - `Currently all IDs must be of the form re.compile('^(?:(?P<namespace>[\\w:-]+)\\/)?(?:(?P<name>[\\w:.-]+?))(?:-v(?P<version>\\d+))?$')`.
2. The parser allows the id even if it's something like `Env-v1.3`, or `Env-v1-v1.2` putting everything into name with `version=None` . This might be an issue when someone registers envs as - `Env-v1.1` , `Env-v1.2`, `Env-v1.3` these all get registered as different names with no version without warning.

RedTachyon commented:
@arjun-kg Thanks, I changed back the error-warning thing during registration. I changed the error message because that's fairly straight-forward and doesn't actually change functionality. Updating the parser would be a whole other mess, so keeping this outside the scope of this

RedTachyon commented:
I think I addressed everything, a few things that I'm deliberately keeping out of scope for now (so that the basic functionality can be merged faster) are:

- The parser regex might be a bit iffy with weirdly written versions
- Plugin system might require an extra warning when failing to import a plugin
- Some tests use a weird construct to emulate `with pytest.raises`, so they might need a revamp
- Add an extra env checker when doing `gym.make` (this would be a completely new feature, so I'd keep it for another PR - for now this just replicates the existing features)

Markus28 commented:
@RedTachyon what about the comment about `check_name_exists`? Why don't we need to filter the specs by namespace?

RedTachyon commented:
@Markus28 I don't think I see the comment?

Markus28 commented:
@RedTachyon Hmm idk, maybe something weird is going on with GitHub, I can't link the comments and it doesn't show up in the review section. These were my comments, I am mostly concerned about the second one:
![image](https://user-images.githubusercontent.com/15806078/164683565-c94fbb30-7a6d-47a0-ba5e-1c5b2518e0c4.png)
![image](https://user-images.githubusercontent.com/15806078/164683609-15a6ecc5-d436-4379-bcda-19e3aa01614b.png)

RedTachyon commented:
@Markus28 that's like the second time I'm seeing the same issue in a week (from another person), you entered these comments as a review and need to click a green button somewhere on top of the page to submit the review for them to actually show up

Markus28 reviewed (commented):
Thanks, forgot that

RedTachyon commented on gym/envs/registration.py:
```diff
@@ -78,558 +75,191 @@ def parse_env_id(id: str) -> Tuple[Optional[str], str, Optional[int]]:
 
+        # Check if the package is installed
+        # If not instruct the user to install the package and then how to instantiate the env
+        if importlib.util.find_spec(relocated_package) is None:
```
So I think this was copy-pasted verbatim from the [previous version of the code](https://github.com/openai/gym/blob/c6deb81ad3b6da7adbbffd4a2df2002a53861d8f/gym/envs/registration.py#L288), it's possible it doesn't work, but that's out of scope here

RedTachyon commented on gym/envs/registration.py:
```diff
@@ -78,558 +75,191 @@ def parse_env_id(id: str) -> Tuple[Optional[str], str, Optional[int]]:
 
-        if self.autoreset:
-            from gym.wrappers.autoreset import AutoResetWrapper
+def check_name_exists(ns: Optional[str], name: str):
```
Good catch, I'll add the `if spec_.namespace == ns` condition

RedTachyon commented on gym/envs/registration.py:
```diff
@@ -680,63 +310,190 @@ def make(id: Literal[
+    """
+    Register an environment with gym. The `id` parameter corresponds to the name of the environment,
+    with the syntax as follows:
+    `(namespace)/(env_name)-(version)`
```
- Updated
- Good point, this actually breaks `with namespace`, I'll fix it
- Yep, I'll fix that too

RedTachyon commented on gym/envs/registration.py:
```diff
@@ -680,63 +310,190 @@ def make(id: Literal[
+        default=None,
+    )
+
+    unversioned_spec = next(
+        (
+            spec_
+            for spec_ in registry.values()
+            if spec_.namespace == spec.namespace
+            and spec_.name == spec.name
+            and spec_.version is None
+        ),
+        None,
+    )
+
+    if unversioned_spec and spec.version is not None:
```
Fixed
